### PR TITLE
Converts Job Hour Requirements System To HOURS Define

### PR DESCRIPTION
**Pull request by san7890**
## About The Pull Request

Hey there,

So, we stored everything associated with "qualifying for a job" in minutes without it being clear that it was supposed to be in minutes aside from a comment and a define. So, I decided to clean up some of the math by using the HOURS and MINUTES macros, and it should be a lot clearer to read and understand,  I hope.

image

That above value is in minutes, and it's based off a DB call. So, this PR contains some certified:tm: bullshit:tm: to respect the values that the DB gives us, without altering the DB-level behavior because that could break it. I've broken plenty in my time here, and I don't particularly want to break this.

This is why we have to do some defines in the if() statements passed through in this PR, to ensure that we properly compare HOURS and MINUTES. I did the math out on a calculator to ensure I wasn't fudging it somehow and I truly do hope we're good.

I went through this a few times over the last few days, and it doesn't seem like I missed any comparisons.

## Why It's Good For The Game

Easier code to read and immediately understand rather than be a bit murked in visibility.

## Changelog

Nothing particularly player facing, at least not until I learn I fucked up somehow.

### CODE REVIEW

File: code/game/objects/items/cards_ids.dm [hunk 1]
```diff
@@ -5,7 +5,7 @@
 #define ID_ICON_BORDERS 1, 9, 32, 24
 
 /// Fallback time if none of the config entries are set for USE_LOW_LIVING_HOUR_INTERN
-#define INTERN_THRESHOLD_FALLBACK_HOURS 15
+#define INTERN_THRESHOLD_FALLBACK_HOURS 15 HOURS
```
Comment (willox) [inline code/game/objects/items/cards_ids.dm:8]: This is no longer in "hours" units and should probably be named something like `INTERN_THRESHOLD_FALLBACK_TIME`

File: code/game/objects/items/cards_ids.dm [hunk 2]
```diff
@@ -935,10 +935,11 @@
...
 		return
 
-	var/intern_threshold = (CONFIG_GET(number/use_low_living_hour_intern_hours) * 60) || (CONFIG_GET(number/use_exp_restrictions_heads_hours) * 60) || INTERN_THRESHOLD_FALLBACK_HOURS * 60
-	var/playtime = user.client.get_exp_living(pure_numeric = TRUE)
+	var/intern_threshold = (CONFIG_GET(number/use_low_living_hour_intern_hours) HOURS) || (CONFIG_GET(number/use_exp_restrictions_heads_hours) HOURS) || INTERN_THRESHOLD_FALLBACK_TIME
+	var/playtime = user.client.get_exp_living(pure_numeric = TRUE) //Pure numeric, so any values returned by this proc will be in minutes.
 
-	if((intern_threshold >= playtime) && (user.mind?.assigned_role.job_flags & JOB_CAN_BE_INTERN))
+	// intern_threshold is in HOURS, while playtime is in MINUTES, so declare the appropriate defines here to allow for proper comparison.
+	if((intern_threshold >= (playtime MINUTES)) && (user.mind?.assigned_role.job_flags & JOB_CAN_BE_INTERN))
```
Comment (willox) [inline code/game/objects/items/cards_ids.dm:942]: The comment states that "intern_threshold is in HOURS" but I think that's incorrect?

Comment (san7890) [reply code/game/objects/items/cards_ids.dm:942]:
```dm
var/intern_threshold = (CONFIG_GET(number/use_low_living_hour_intern_hours) HOURS) || (CONFIG_GET(number/use_exp_restrictions_heads_hours) HOURS) || INTERN_THRESHOLD_FALLBACK_TIME
```

intern_threshold will pass via the decisecond define for HOURS (to respect existing config). So, it should be a correct comment? It's meant to cross-compare at the decisecond level though with the changes I'm passing through. I'll see if I can fix the comments.

File: code/modules/jobs/job_exp.dm [hunk 3]
```diff
@@ -17,20 +18,20 @@ GLOBAL_PROTECT(exp_to_update)
 	var/isexempt = C.prefs.db_flags & DB_FLAG_EXEMPT
 	if(isexempt)
 		return 0
-	var/my_exp = C.calc_exp_type(get_exp_req_type())
-	var/job_requirement = get_exp_req_amount()
-	if(my_exp >= job_requirement)
+	var/my_exp = C.calc_exp_type(get_exp_req_type()) // this value is returned in minutes.
+	var/job_requirement = get_exp_req_amount() // this value is returned in hours.
```
Comment (willox) [inline code/modules/jobs/job_exp.dm:22]: "this value is returned in hours"

it's deciseconds, right?

Comment (san7890) [reply code/modules/jobs/job_exp.dm:22]: *yells into pillow*

you're right

File: code/game/objects/items/cards_ids.dm [hunk 4]
```diff
@@ -935,10 +935,11 @@
 	if(!SSdbcore.Connect())
 		return
 
-	var/intern_threshold = (CONFIG_GET(number/use_low_living_hour_intern_hours) * 60) || (CONFIG_GET(number/use_exp_restrictions_heads_hours) * 60) || INTERN_THRESHOLD_FALLBACK_HOURS * 60
-	var/playtime = user.client.get_exp_living(pure_numeric = TRUE)
+	var/intern_threshold = (CONFIG_GET(number/use_low_living_hour_intern_hours) HOURS) || (CONFIG_GET(number/use_exp_restrictions_heads_hours) HOURS) || INTERN_THRESHOLD_FALLBACK_TIME
```
Comment (Watermelon914) [inline code/game/objects/items/cards_ids.dm:938]: do `* (1 HOURS)`, don't do `var HOURS`

**Review by Watermelon914 — CHANGES_REQUESTED** [review]

File: code/game/objects/items/cards_ids.dm [hunk 5]
```diff
@@ -5,7 +5,7 @@
 #define ID_ICON_BORDERS 1, 9, 32, 24
 
 /// Fallback time if none of the config entries are set for USE_LOW_LIVING_HOUR_INTERN
-#define INTERN_THRESHOLD_FALLBACK_HOURS 15
+#define INTERN_THRESHOLD_FALLBACK_TIME 15 HOURS
```
Comment (JohnFulpWillard) [inline code/game/objects/items/cards_ids.dm:8]:
```suggestion
#define INTERN_THRESHOLD_FALLBACK_TIME (15 HOURS)
```

File: code/modules/jobs/job_exp.dm [hunk 6]
```diff
@@ -17,20 +18,20 @@ GLOBAL_PROTECT(exp_to_update)
...
-	var/my_exp = C.calc_exp_type(get_exp_req_type())
-	var/job_requirement = get_exp_req_amount()
-	if(my_exp >= job_requirement)
+	var/my_exp = C.calc_exp_type(get_exp_req_type()) // this value is returned in minutes via the DB.
+	var/job_requirement = get_exp_req_amount() // this value is returned in deciseconds.
+	if((my_exp MINUTES) >= job_requirement) // The evaluation done here is done on the deciseconds level using the time defines.
 		return 0
 	else
-		return (job_requirement - my_exp)
+		return (job_requirement - my_exp MINUTES)
```
Comment (optimumtact) [inline code/modules/jobs/job_exp.dm:26]: suspect you need brackets here (my_exp MINUTES)

**Pull request merged**